GPU/Common/GPUStateUtils.cpp

```diff
@@ -386,3 +408,627 @@ LogicOpReplaceType ReplaceLogicOpType() {
+	// This is only useful for INCR/DECR/INVERT.  Others can write directly.
+	switch (stencilType) {
+	case STENCIL_VALUE_INCR_4:
+	case STENCIL_VALUE_INCR_8:
+		// We'll add the incremented value output by the shader.
+		blendState.enabled = true;
+		blendState.setFactors(srcBlend, dstBlend, BlendFactor::ONE, BlendFactor::ONE);
+		blendState.setEquation(blendEq, BlendEq::ADD);
+		break;
+
+	case STENCIL_VALUE_DECR_4:
+	case STENCIL_VALUE_DECR_8:
+		// We'll subtract the incremented value output by the shader.
+		blendState.enabled = true;
+		blendState.setFactors(srcBlend, dstBlend, BlendFactor::ONE, BlendFactor::ONE);
+		blendState.setEquation(blendEq, BlendEq::SUBTRACT);
+		break;
+
+	case STENCIL_VALUE_INVERT:
+		// The shader will output one, and reverse subtracting will essentially invert.
+		blendState.enabled = true;
+		blendState.setFactors(srcBlend, dstBlend, BlendFactor::ONE, BlendFactor::ONE);
+		blendState.setEquation(blendEq, BlendEq::REVERSE_SUBTRACT);
+		break;
+
+	default:
+		if (srcBlend == BlendFactor::ONE && dstBlend == BlendFactor::ZERO && blendEq == BlendEq::ADD) {
+			blendState.enabled = false;
+		} else {
+			blendState.enabled = true;
+			blendState.setFactors(srcBlend, dstBlend, BlendFactor::ONE, BlendFactor::ZERO);
+			blendState.setEquation(blendEq, BlendEq::ADD);
+		}
+		break;
+	}
+}
+
+// Called even if AlphaBlendEnable == false - it also deals with stencil-related blend state.
+
+void ConvertBlendState(GenericBlendState &blendState) {
+	// Blending is a bit complex to emulate.  This is due to several reasons:
+	//
+	//  * Doubled blend modes (src, dst, inversed) aren't supported in OpenGL.
+	//    If possible, we double the src color or src alpha in the shader to account for these.
+	//    These may clip incorrectly, so we avoid unfortunately.
+	//  * OpenGL only has one arbitrary fixed color.  We premultiply the other in the shader.
+	//  * The written output alpha should actually be the stencil value.  Alpha is not written.
+	//
+	// If we can't apply blending, we make a copy of the framebuffer and do it manually.
+	gstate_c.allowShaderBlend = !g_Config.bDisableSlowFramebufEffects;
+
+	blendState.applyShaderBlending = false;
+	blendState.dirtyShaderBlend = false;
+	blendState.useBlendColor = false;
+
+	ReplaceBlendType replaceBlend = ReplaceBlendWithShader(gstate_c.allowShaderBlend, gstate.FrameBufFormat());
+	ReplaceAlphaType replaceAlphaWithStencil = ReplaceAlphaWithStencil(replaceBlend);
+	bool usePreSrc = false;
+
+	switch (replaceBlend) {
+	case REPLACE_BLEND_NO:
+		blendState.resetShaderBlending = true;
+		// We may still want to do something about stencil -> alpha.
+		ApplyStencilReplaceAndLogicOp(replaceAlphaWithStencil, blendState);
+		return;
+
+	case REPLACE_BLEND_COPY_FBO:
+		blendState.applyShaderBlending = true;
+		blendState.resetShaderBlending = false;
+		blendState.replaceAlphaWithStencil = replaceAlphaWithStencil;
+		break;  // Surely this should be return??
```

Comment:
Just noting here again, break makes sense because we might not apply the copy fbo effect.  Honestly, it might be better to fall back to PRE_SRC or etc. in that case... but it shouldn't be common.

-[Unknown]